src/Raven.Client/Documents/Session/AsyncDocumentQuery.cs

```diff
@@ -1011,5 +1082,25 @@ public IAsyncDocumentQuery<TResult> AsyncQuery<TResult>(string indexName, string
 
             return CreateDocumentQueryInternal<TResult>();
         }
+
+        public IAsyncGraphQuery<T> With<TOther>(string alias, IRavenQueryable<TOther> query, CancellationToken token = default)
+        {
+            //TODO:Find a proper way of do this extraction
+            var queryInspector = (RavenQueryInspector<TOther>)query;
+            var docQuery = (AsyncDocumentQuery<TOther>)queryInspector.GetAsyncDocumentQuery(x => x.ParameterPrefix = $"w{WithTokens.Count}p");
+            if (docQuery.SelectTokens?.Count > 0)
+            {
+                throw new NotSupportedException($"Select is not premitted in a 'With' clause in query:{docQuery}");
+            }
+
+
+            foreach (var keyValue in docQuery.QueryParameters)
+            {
+                QueryParameters.Add(keyValue.Key, keyValue.Value);
```

Comment:
Maybe we should use `AddParameter` method. I do not remember if we are doing some validation there or not.

Reply:
I need to set the key of the parameter in the "parent" query the AddParameter only gets the value of the parameter.
The keys are promised to be unique so this will not throw (since we never remove parameters from the dictionary and I use the count of values and with statements)